### ISSUE TITLE
How to reliably detect transcoded JPEG?

### ISSUE THREAD

**Issue by natewind** [libjxl/libjxl#3252]
When I inspect a transcoded image with `jxlinfo`, it says `JPEG bitstream reconstruction data available`. Can I rely on that exact string always being there and match against it? If not, it there a more formal way to check whether a JXL is a transcoded JPEG?

**Comment by jonsneyers** [comment]
If that string is not there, then libjxl will not be able to reconstruct the JPEG.
You can have a transcoded JPEG where the reconstruction data is not available though. In that case the image data is still fine, but you cannot reconstruct the exact original JPEG file.

**Comment by natewind** [comment]
No, I get that, it’s not the point. Can I rely on the exact wording staying the same? It’s not a number or a boolean but an informal phrase, so I’m a little unsure as to whether I can reliably match against it (in a Bash function).

BTW, what I’m trying to do is to automatically convert to JPEG if it can be reconstructed, or to PNG otherwise — so I always choose the smallest lossless option. Maybe there is a more straightforward way to do it?

**Comment by jonsneyers** [comment]
Checking for that string should work.

**Comment by natewind** [comment]
So it won’t change in future versions?

**Comment by alistair7** [comment]
I was reluctant to rely on `jxlinfo | grep` too, so I wrote a minimal Python script for detecting reconstructable JPEGs by parsing the JXL container and looking for a `jbrd` box.

If it's any use to you:
https://gist.github.com/alistair7/3dfbc021db19a9ff237f11a2c3561056

Perhaps it would be useful to add an `--is_lossless_jpeg` option to `jxlinfo` for explicitly querying this.  (Or, at the risk of completely over-engineering the solution, an option to output the info in a machine-parseable or custom format.)

**Comment by natewind** [comment]
> I was reluctant to rely on `jxlinfo | grep` too, so I wrote a minimal Python script for detecting reconstructable JPEGs by parsing the JXL container and looking for a `jbrd` box.
> 
> If it's any use to you: https://gist.github.com/alistair7/3dfbc021db19a9ff237f11a2c3561056

Somebody on Reddit suggested an equivalent Bash line, I just haven’t tried it yet:

```sh
head -c 64 test.jxl | grep -ao 'jbrd' | wc -l
```

Which would be more convenient than mixing in Pyton. But thanks for the suggestion!

> Perhaps it would be useful to add an `--is_lossless_jpeg` option to `jxlinfo` for explicitly querying this. (Or, at the risk of completely over-engineering the solution, an option to output the info in a machine-parseable or custom format.)

+++

**Comment by natewind** [comment]
Here’s the full solution in case anyone else finds it useful:

```sh
error() { echo -e "\033[1;31mError:\033[0m $@" 1>&2; }

jxl()
{
	if [[ $# -ne 1 ]]; then
		error "Wrong number of arguments!"
		return 1
	elif [[ ! -e "$1" ]]; then
		error "$1 doesn’t exist!"
		return 1
	fi

	local type=$(file -b --mime-type "$1")
	local out="${1%.*}.jxl"

	if [[ $type == "image/png" ]]; then
		cjxl -q 100 "$1" "$out"
	elif [[ $type == "image/jpeg" ]]; then
		cjxl --lossless_jpeg=1 "$1" "$out"
	else
		error "$1 is not PNG or JPEG!"
		return 1
	fi
}

unjxl()
{
	if [[ $# -ne 1 ]]; then
		error "Wrong number of arguments!"
		return 1
	elif [[ $(file -b --mime-type "$1") != "image/jxl" ]]; then
		error "$1 is not JPEG XL!"
		return 1
	fi

	if head -c 64 "$1" | grep -qa "jbrd"; then
		local ext="jpg"
	else
		local ext="png"
	fi

	djxl "$1" "${1%.*}.$ext"
}
```

**Comment by alistair7** [comment]
> if head -c 64 test.jxl | grep -qa "jbrd"; then ...

That's definitely easier!  But bear in mind the `jbrd` doesn't have to be in the first 64 bytes of the file (e.g., I have some files where the `jbrd` comes after several KB of Exif data - although the latest cjxl now seems to prefer to put jbrd first), so it's not 100% reliable.

It would also give a false positive if the file just coincidentally contained the sequence of bytes `j` `b` `r` and `d` somewhere near the start.

If the odd false positive/negative isn't a disaster, grepping for `jbrd` definitely wins on simplicity, and might be good enough :).  But for a truly reliable method I think the only options are using the libjxl API or parsing the ISO box structure "by hand" (like the Python script).

**Comment by natewind** [comment]
Damn, then it’s not good enough...

**Comment by natewind** [comment]
I guess the only sane solution would be this:

> Perhaps it would be useful to add an `--is_lossless_jpeg` option to `jxlinfo` for explicitly querying this. (Or, at the risk of completely over-engineering the solution, an option to output the info in a machine-parseable or custom format.)

So could someone please add an `enhancement` label?

**Comment by natewind** [comment]
BTW, here’s an alternative approach that relies on `jxlinfo`. It matches the entire output, so if the format changes, it will fail instead of silently doing the wrong thing.

```sh
jxl_jpeg='^JPEG XL file format container \(ISO/IEC 18181-2\)
JPEG XL image, [0-9]+x[0-9]+, \(possibly\) lossless, [0-9]+-bit RGB
Color space: RGB, D65, sRGB primaries, sRGB transfer function, rendering intent: Relative
JPEG bitstream reconstruction data available$'

jxl_png='^JPEG XL image, [0-9]+x[0-9]+, \(possibly\) lossless, [0-9]+-bit RGB\+Alpha
Color space: RGB, D65, sRGB primaries, sRGB transfer function, rendering intent: Relative$'

unjxl()
{
	...

	local info=$(jxlinfo "$1")

	if [[ "$info" =~ $jxl_jpeg ]]; then
		local ext="jpg"
	elif [[ "$info" =~ $jxl_png ]]; then
		local ext="png"
	else
		error "Unknown jxlinfo format!"
		return 1
	fi

	...
}
```

**Comment by Traneptora** [comment]
One option is to call `djxl input.jxl output.jpg --pixels_to_jpeg=0` which will fail if it can't be done.

**Comment by natewind** [comment]
> One option is to call `djxl input.jxl output.jpg --pixels_to_jpeg=0` which will fail if it can't be done.

`--pixels_to_jpeg` doesn’t take arguments.
